files/image_config/hostcfgd/hostcfgd

```diff
@@ -0,0 +1,339 @@
+                self.tacplus_servers.append(server)
+
+        self.tacplus_servers = sorted(self.tacplus_servers,
+                                      key=lambda server: server.priority,
+                                      reverse=True)
+        if modify_conf:
+            self.modify_pam_auth_file()
+            self.modify_nss_conf()
+
+    def get_pam_auth(self, pam_module):
+        if pam_module == AAA_LOCAL:
+            pam_auth = PAM_LOCAL_MODULE
+            if self.debug:
+                pam_auth += " debug"
+            pam_auth += " nullok try_first_pass\n"
+            return pam_auth
+        else:
+            return pam_module.get_pam_conf(self.tacplus_global, self.debug)
+
+    def modify_pam_auth_file(self):
+        if self.auth_failthrough:
+            pam_control = "[success=done new_authtok_reqd=done default=ignore]"
+        else:
+            pam_control = "[success=done new_authtok_reqd=done default=ignore auth_err=die]"
+
+        pam_modules = []
+        auth_file_body = ""
+
+        # Set local and tacacs+ pam priority
+        # ['local'] or ['local','tacacs+'] or ['tacacs+', 'local'] or ['tacacs+']
+        if self.auth_login == [AAA_LOCAL]:
+            pam_modules = [AAA_LOCAL]
+        elif self.auth_login == [AAA_LOCAL, AAA_TACPLUS]:
+            pam_modules = [AAA_LOCAL] + self.tacplus_servers
+        elif self.auth_login == [AAA_TACPLUS] or self.auth_login == [AAA_TACPLUS, AAA_LOCAL]:
+            # Don't accept only TACACS+ authentication
+            # Make sure root will always authentication on local, not TACACS+
+            pam_modules = self.tacplus_servers + [AAA_LOCAL]
+            auth_file_body += "auth\t[success=%d new_authtok_reqd=done default=ignore]\t" % (len(pam_modules)-1)
+            auth_file_body += "pam_succeed_if.so user = root debug\n"
+
+        if len(pam_modules):
+            for pam_module in pam_modules[:-1]:
+                line = self.get_pam_auth(pam_module)
+                if line is not "":
+                    auth_file_body += "auth\t" + pam_control + "\t" + line
+
+            pam_module = pam_modules[-1]
+            line = self.get_pam_auth(pam_module)
+            auth_file_body += "auth\t[success=1 default=ignore]\t" + line
+
+        with open(PAM_AUTH_FILE, "w") as f:
+            f.write(AUTH_FILE_HEADER + auth_file_body + AUTH_FILE_FOOTER)
+
+        # Modify common-auth include file in /etc/pam.d/type and sshd
+        if os.path.isfile(PAM_AUTH_FILE):
+            os.system("sed -i -e '/^@include/s/common-auth$/common-auth-sonic/' /etc/pam.d/sshd")
+            os.system("sed -i -e '/^@include/s/common-auth$/common-auth-sonic/' /etc/pam.d/login")
+        else:
+            os.system("sed -i -e '/^@include/s/common-auth-sonic$/common-auth/' /etc/pam.d/sshd")
+            os.system("sed -i -e '/^@include/s/common-auth-sonic$/common-auth/' /etc/pam.d/login")
+
+    # Set tacacs+ server in nss-tacplus conf
+    def modify_nss_conf(self):
```

Comment:
it is very difficult to know what the actual conf file generated, it is better to design a template and use jinja2 library in python to generate the conf file. You can refer to the code in sonic-cfggen.

Reply:
I have changed it with jinja2 template.